clients/hadoopfs/src/main/java/io/lakefs/MetadataClient.java

```diff
@@ -0,0 +1,89 @@
+package io.lakefs;
+
+import com.amazonaws.services.s3.model.GetObjectMetadataRequest;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.beans.IntrospectionException;
+import java.beans.Introspector;
+import java.beans.PropertyDescriptor;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.URI;
+
+public class MetadataClient {
+    public static final Logger LOG = LoggerFactory.getLogger(MetadataClient.class);
+    private final FileSystem fs;
+    private Method getMetadata;
+    private Object s3client;
+
+    public MetadataClient(FileSystem fs) {
+        this.fs = fs;
+        if (this.fs == null) {
+            return;
+        }
+        try {
+            Method amazonS3ClientGetter = fs.getClass().getDeclaredMethod("getAmazonS3Client");
+            amazonS3ClientGetter.setAccessible(true);
+            this.s3client = amazonS3ClientGetter.invoke(fs);
+            this.getMetadata = s3client.getClass().getDeclaredMethod("getObjectMetadata", GetObjectMetadataRequest.class);
+        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+            LOG.warn("get underlying get object metadata method:", e);
+        }
+    }
+
+    /**
+     * get object metadata by physical address
+     * @param physicalUri physical uri of object
+     * @return ObjectMetadata filled with Etag and content length
+     * @throws IOException
+     */
+    ObjectMetadata getObjectMetadata(URI physicalUri) throws IOException {
+        String bucket = physicalUri.getHost();
+        String key = physicalUri.getPath().substring(1);
+        if (this.s3client != null && this.getMetadata != null) {
```

Comment:
I don't understand.  When is this method better than stealing the file status from S3A?  This will not be more efficient than `getFileStatus` on the network... but it might be less one day if we can tell S3A to cache things.  And it preserves a dependency on AWS code, which anyway changed version significantly since the Hadoop-AWS 2.7.7 days.  And we anyway need the stolen ETag from S3AFIleStatus method.  Can we drop this version and keep only that one?

Reply:
At first, there was just 'getFileStatus' and extract 'etag' - everything worked, except on old Spark cluster 2.4.x where the S3AFileStatus didn't contain that value.

Switched to extracting the s3 client from the underlying file system, as a fallback. But, it caused two calls one for the FileStatus and one for getObjectMetadata on the s3 client in case you are on the older version.
I understand the issue that we have when we perform s3 call on-call version.
Just to be clear that the older version will work a little harder in order for us to not use the s3 client in the newer version.